graylog2-web-interface/src/components/visualizations/GraphVisualization.jsx

```diff
@@ -154,6 +160,87 @@ const GraphVisualization = React.createClass({
     } else {
       this.graph.redraw();
     }
+    if (this.props.config.threshold) {
+      this.renderThreshold();
+    }
+  },
+  // Draws a horizontal threshold line in the graph
+  renderThreshold() {
+    const threshold = this.props.config.threshold;
+    const thresholdColor = this.props.config.threshold_color || '#f00';
+    const thresholdTooltip = this.props.config.threshold_tooltip || `threshold: ${threshold}`;
+    const thresholdDotHidden = 1e-6;
+    const thresholdDotVisible = 0.8;
+    const thresholdDotRadius = 4;
+
+    this.graph.on('renderlet.threshold', (chart) => {
+      const lineData = [
+        {
+          x: chart.x().range()[0],
+          y: chart.y()(threshold),
+        },
+        {
+          x: chart.x().range()[1],
+          y: chart.y()(threshold),
+        },
+      ];
+
+      const line = d3.svg.line()
+        .x(d => d.x)
+        .y(d => d.y)
+        .interpolate('linear');
+
+      const chartBody = chart.select('g.chart-body');
+      const paths = chartBody.selectAll('path.threshold').data([lineData]);
+      paths // Modify the existing path
+        .attr('stroke', thresholdColor)
+        .attr('d', line);
+      paths.enter() // This will only do something if there isn't a path yet
+        .append('path')
+        .attr('class', 'threshold')
+        .attr('stroke', thresholdColor)
+        .attr('stroke-width', 1)
+        .attr('stroke-dasharray', ('2', '2'))
+        .attr('d', line);
+      paths.exit().remove(); // Remove any outdated paths
+
+      // Collect all x-axis values
+      const xValues = chart.data().reduce((list, value) => {
+        value.values.forEach(v => list.push(v.x));
+        return list;
+      }, []);
+
+      // Put circles on the threshold line to make it easier to show the tooltip
+      const dots = chartBody.selectAll('circle.threshold').data(xValues);
+      dots.enter()
+        .append('circle')
+        .attr('class', 'threshold')
+        .attr('r', thresholdDotRadius)
+        .attr('rel', 'tooltip') // To make the tooltip helper pick it up
+        .attr('data-original-title', () => {
+          return `<div class="datapoint-info">${thresholdTooltip}</div>`;
```

Comment:
Will `thresholdTooltip` ever have some user defined value? Here we use d3 directly and are away from React's protection against code injection.

Reply:
Very good catch. Escaping this one.